### PR TITLE
STM32MP1 : Clock_control and GPIO support

### PR DESCRIPTION
**Pull request by yboutreux**
Add compatibility of stm32mp157c_dk2 board with clock_control and GPIO driver.
Also fix and cleanup of some files related to the stm32mp157c_dk2 or the STM32 family.

### CODE REVIEW

**Comment by erwango** [comment]
@arnop2

**Comment by arnop2** [comment]
For information this pull request the follow https://github.com/zephyrproject-rtos/zephyr/pull/13794  pull request.

File: boards/arm/stm32mp157c_dk2/doc/stm32mp157_dk2.rst [hunk 1]
```diff
@@ -0,0 +1,302 @@
...
+The Zephyr console output is assigned by default to the ram console to be dumped
+by the Linux Remoteproc Framework on Cortex®-A7 core. The UART 4 can enabled as
+Cortex®-M4 console.
+
+Programming and Debugging
+*************************
+The STM32MP157C doesn't have QSPI flash for the Cortex®-M4  and it needs to be
+started by the Cortex®-A7 core. The Cortex®-A7 core is responsible to load the
+Cortex®-M4 binary application into the RAM, get the Cortex®-M4 out of reset.
+The Cortex®-A7 can perform these steps at bootloader level or after the Linux
```
Comment (dbkinder) [inline boards/arm/stm32mp157c_dk2/doc/stm32mp157_dk2.rst]: Linux?  I suspect this should say ``Zephyr``, right?

File: boards/arm/stm32mp157c_dk2/doc/stm32mp157_dk2.rst [hunk 2]
```diff
@@ -0,0 +1,302 @@
...
+system has booted.
+
+The Cortex®-M4 can use up to 2 different RAMs. The program pointer starts at
+address 0x00000000 (RETRAM), the vector table should be loaded at this address
+These are the memory mapping for Cortex®-A7 and Cortex®-M4:
```
Comment (dbkinder) [inline boards/arm/stm32mp157c_dk2/doc/stm32mp157_dk2.rst]:
```suggestion
These are the memory mappings for Cortex®-A7 and Cortex®-M4:
```

File: boards/arm/stm32mp157c_dk2/doc/stm32mp157_dk2.rst [hunk 3]
```diff
@@ -0,0 +1,302 @@
...
++============+=======================+========================+======================+
+| RETRAM     | 0x38000000-0x3800FFFF | 0x00000000-0x0000FFFF  | 64KB                 |
++------------+-----------------------+------------------------+----------------------+
+| MCUSRAM    | 0x10000000-0x1005FFFF | 0x10000000-0x1005FFFF  | 384KB                |
++------------+-----------------------+------------------------+----------------------+
+| DDR        | 0xC0000000-0xFFFFFFFF |                        | up to 1 GB           |
++------------+-----------------------+------------------------+----------------------+
+
+
+Please refer to `stm32mp157c boot Cortex-M4 firmware`_ wiki page for instruction to load and
```
Comment (dbkinder) [inline boards/arm/stm32mp157c_dk2/doc/stm32mp157_dk2.rst]:
```suggestion
Refer to the `stm32mp157c boot Cortex-M4 firmware`_ wiki page for instruction to load and
```

File: boards/arm/stm32mp157c_dk2/doc/stm32mp157_dk2.rst [hunk 4]
```diff
@@ -0,0 +1,302 @@
...
+start the Cortex-M4 firmware.
+
+Debugging
+=========
+
+You can debug an application using OpenOCD and GDB. The Solution proposed below
+is based on the Linux STM32MP1 SDK OpenOCD and is available only for Linux environment.
```
Comment (dbkinder) [inline boards/arm/stm32mp157c_dk2/doc/stm32mp157_dk2.rst]:
```suggestion
is based on the Linux STM32MP1 SDK OpenOCD and is available only for a Linux development environment.
```

File: boards/arm/stm32mp157c_dk2/doc/stm32mp157_dk2.rst [hunk 5]
```diff
@@ -0,0 +1,302 @@
...
+The firmware have first to be loaded by Cortex®-A7. Then to debug the zephyr
```
Comment (dbkinder) [inline boards/arm/stm32mp157c_dk2/doc/stm32mp157_dk2.rst]: I'm not sure what "attach on it" means, and this rewrite might not be technically correct:
```
The firmware must first be loaded by the Cortex®-A7. The board must be attached
to your development system to debug the Zephyr application using OpenOCD
via the serial port.
```

**Review by dbkinder — CHANGES_REQUESTED** [review]

File: boards/arm/stm32mp157c_dk2/doc/stm32mp157_dk2.rst [hunk 6]
```diff
@@ -0,0 +1,302 @@
...
+############################
+
+Overview
+********
+
+The STM32MP157-DK2 Discovery board leverages the capacities of the STM32MP157
+multi-core processor,composed of a dual Cortex®-A7 and a single Cortex®-M4 core.
+Zephyr OS is ported to run on the Cortex®-M4 core.
+
+  - Common features:
```
[the remark this answers is not in the document]
Comment (arnop2) [reply boards/arm/stm32mp157c_dk2/doc/stm32mp157_dk2.rst]: @dbkinder  this file is part of the  #13794 initial pull request, you already reviewed. As it is still waiting the merge  windows,  i will update the rst file in the initial pool request to take your remarks into account. it is ok for you, to  post your next messages in #13794 for the rst file?

File: boards/arm/stm32mp157c_dk2/doc/stm32mp157_dk2.rst [hunk 7]
```diff
@@ -0,0 +1,302 @@
...
+  - AES 128, 192, 256, TDES
+  - HASH (MD5, SHA-1, SHA224, SHA256), HMAC
+  - 2 × true random number generator (3 oscillators each)
+  - 2 × CRC calculation unit
+
+- Debug mode:
+
+  - Arm® CoreSight™ trace and debug: SWD and JTAG interfaces
+  - 8-Kbyte embedded trace buffer
+  - 3072-bit fuses including 96-bit unique ID, up to 1184-bit available for user
```
[the remark this answers is not in the document]
Comment (arnop2) [reply boards/arm/stm32mp157c_dk2/doc/stm32mp157_dk2.rst]: The issue here is that main of this hardware is also accessible from the cortex-M4. So could be related to the Zephyr. I will suppress some details to keep only main features

File: boards/arm/stm32mp157c_dk2/doc/stm32mp157_dk2.rst (hunk 1 above)
Comment (arnop2) [reply boards/arm/stm32mp157c_dk2/doc/stm32mp157_dk2.rst]: No it is Linux. The Cortex-M4 firmware is not independent. It has to be loaded by the Cortex-A7. 
In the ST distribution an Uboot and a Linux is running on Cortex-A7. So the Zephyr firmware is either loaded by Uboot or by the Linux.

File: boards/arm/stm32mp157c_dk2/doc/stm32mp157_dk2.rst (hunk 5 above)
Comment (arnop2) [reply boards/arm/stm32mp157c_dk2/doc/stm32mp157_dk2.rst]: Attach means here that developer can connect a debugger on the fly to debug the software that as been already loaded and is running. As there is some interactions between the Zephyr and the Linux ( e.g rpmsg) this mode is privileged to not reset the Cortex-M4 with the debugger.

Proposal: `The firmware must first be loaded by the Cortex®-A7. Then developer can attach the debugger to the running Zephyr using OpenOCD.`

**Comment by yboutreux** [comment]
@dbkinder commits from @arnop2 https://github.com/zephyrproject-rtos/zephyr/pull/13794 has been approuved and merged, so those commits are not here anymore, can you update your review ? Thanks

**Review by erwango — APPROVED** [review]

File: boards/arm/stm32mp157c_dk2/doc/stm32mp157_dk2.rst (hunk 6 above)
[the remark this answers is not in the document]
Comment (dbkinder) [reply boards/arm/stm32mp157c_dk2/doc/stm32mp157_dk2.rst]: Oh, ok.

**Review by dbkinder — APPROVED** [review]

**Comment by yboutreux** [comment]
Thanks for the reviews and the merge !

**Pull request merged**